app/uk/gov/hmrc/essttpstubs/controllers/PegaController.scala

```diff
@@ -16,37 +16,89 @@
 
 package uk.gov.hmrc.essttpstubs.controllers
 
+import cats.implicits.catsSyntaxEq
 import play.api.Logger
-import play.api.libs.json.Json
-import play.api.mvc.{Action, AnyContent, ControllerComponents}
+import play.api.libs.json.{JsValue, Json}
+import play.api.mvc._
+import uk.gov.hmrc.essttpstubs.config.AppConfig
+import uk.gov.hmrc.essttpstubs.model.PegaOauthToken
+import uk.gov.hmrc.essttpstubs.repo.PegaTokenRepo
 import uk.gov.hmrc.play.bootstrap.backend.controller.BackendController
 
+import java.time.{LocalDateTime, ZoneOffset}
 import javax.inject.{Inject, Singleton}
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.{ExecutionContext, Future}
 import scala.util.Random
 
 @Singleton
-class PegaController @Inject() (cc: ControllerComponents) extends BackendController(cc) {
+class PegaController @Inject() (cc: ControllerComponents, repo: PegaTokenRepo, appConfig: AppConfig)(implicit ec: ExecutionContext) extends BackendController(cc) {
 
   val logger: Logger = Logger(this.getClass)
 
-  val token: Action[AnyContent] = Action { implicit request =>
-    val response = Json.parse(
+  private val pegaTokenExpiryTime: FiniteDuration = appConfig.pegaTokenExpiryTime
+
+  val token: Action[AnyContent] = Action.async { implicit request =>
+
+    repo.findPegaToken().flatMap {
+      case None =>
+        generateAndSaveToken()
+
+      case Some(pegaToken) =>
+        val expirationTime = pegaToken.validFrom.plusSeconds(pegaTokenExpiryTime.toSeconds)
+        if (LocalDateTime.now().isAfter(expirationTime)) {
+          generateAndSaveToken()
+        } else {
+          val remainingTime = expirationTime.toEpochSecond(ZoneOffset.UTC) - LocalDateTime.now().toEpochSecond(ZoneOffset.UTC)
+          val response = generateTokenResponse(pegaToken.accessToken, remainingTime.toInt)
```

Comment:
if you're multiplying the remaining time by two in the cases above then we should probably do it here as well. It'd be worth sticking in a comment to say we're multiplying by two to be able to simulate the 401 expired token scenario so we know why there's a `*2` in the stub